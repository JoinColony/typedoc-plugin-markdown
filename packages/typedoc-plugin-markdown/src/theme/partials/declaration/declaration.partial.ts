import {
  DeclarationReflection,
  LiteralType,
  ReflectionKind,
  ReflectionType,
} from 'typedoc';
import { MarkdownThemeRenderContext } from '../../theme.context';
import {
  escapeChars,
  heading,
  stripComments,
  stripLineBreaks,
} from '../../theme.utils';

export const declarationPartial = (
  context: MarkdownThemeRenderContext,
  props: DeclarationReflection,
) => {
  return [
    declarationTitle(context, props),
    declarationBody(context, props),
  ].join('\n\n');
};

const declarationTitle = (
  context: MarkdownThemeRenderContext,
  props: DeclarationReflection,
) => {
  const md: string[] = [];

  if (props.flags && props.flags.length > 0 && !props.flags.isRest) {
    md.push(' ' + props.flags.map((flag) => `\`${flag}\``).join(' '));
  }
  md.push(`${props.flags.isRest ? '... ' : ''} **${escapeChars(props.name)}**`);
  if (props instanceof DeclarationReflection && props.typeParameters) {
    md.push(
      `<${props.typeParameters
        .map((typeParameter) => `\`${typeParameter.name}\``)
        .join(', ')}\\>`,
    );
  }

  if (!props.parent?.kindOf(ReflectionKind.Enum)) {
    md.push(getDeclarationType(context, props));
  }

  if (
    !(props.type instanceof LiteralType) &&
    props.defaultValue &&
    props.defaultValue !== '...'
  ) {
    md.push(` = \`${stripLineBreaks(stripComments(props.defaultValue))}\``);
  }
  return md.join('');
};

const declarationBody = (
  context: MarkdownThemeRenderContext,
  props: DeclarationReflection,
) => {
  const md: string[] = [];

  const typeDeclaration = (props.type as any)
    ?.declaration as DeclarationReflection;

  if (props.comment) {
    md.push(context.commentPartial(props.comment));
  }

  if (props.typeParameters) {
    md.push(heading(4, 'Type parameters'));
    md.push(context.typeParameterTablePartial(props.typeParameters));
  }

  if (typeDeclaration?.indexSignature) {
    md.push(context.indexSignaturePartial(typeDeclaration.indexSignature));
  }

  if (typeDeclaration?.signatures) {
    md.push(
      heading(
        4,
        typeDeclaration.children ? 'Call signature' : 'Type declaration',
      ),
    );
    typeDeclaration.signatures.forEach((signature) => {
      md.push(context.signaturePartial(signature, true));
    });
  }

  if (typeDeclaration?.children) {
    md.push(heading(4, 'Type declaration'));
    md.push(context.propertyTablePartial(typeDeclaration.children));
  }

  md.push(context.sourcesPartial(props));

  return md.join('\n\n');
};

const getDeclarationType = (
  context: MarkdownThemeRenderContext,
  reflection: DeclarationReflection,
) => {
  const reflectionType = reflection.type as ReflectionType;
  if (reflectionType && reflectionType.declaration?.children) {
    return ': `Object`';
  }
  return (
    ': ' +
    context.typePartial(reflectionType ? reflectionType : reflection, 'object')
  );
};
