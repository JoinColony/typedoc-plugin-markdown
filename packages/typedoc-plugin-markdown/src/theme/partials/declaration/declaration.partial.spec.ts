import { DeclarationReflection, ProjectReflection } from 'typedoc';
import { declarationPartial } from './declaration.partial';

describe(`Declaration Partial:`, () => {
  let project: ProjectReflection;

  beforeAll(async () => {
    project = global.getProject(__dirname);
  });

  test(`should compile a const with default value`, () => {
    expect(
      declarationPartial(
        global.getMockContext(),
        project.findReflectionByName(
          'stringConstWithDefaultValue',
        ) as DeclarationReflection,
      ),
    ).toMatchSnapshot();
  });

  test(`should compile an undefined declaration`, () => {
    expect(
      declarationPartial(
        global.getMockContext(),
        project.findReflectionByName(
          'undefinedNumberDeclaration',
        ) as DeclarationReflection,
      ),
    ).toMatchSnapshot();
  });

  test(`should compile object literal declaration`, () => {
    expect(
      declarationPartial(
        global.getMockContext(),
        project.findReflectionByName(
          'objectLiteralDeclaration',
        ) as DeclarationReflection,
      ),
    ).toMatchSnapshot();
  });

  test(`should compile object literal cast as a const`, () => {
    expect(
      declarationPartial(
        global.getMockContext(),
        project.findReflectionByName(
          'objectLiteralAsConstDeclaration',
        ) as DeclarationReflection,
      ),
    ).toMatchSnapshot();
  });

  test(`should compile type literal declaration`, () => {
    expect(
      declarationPartial(
        global.getMockContext(),
        project.findReflectionByName(
          'typeLiteralDeclaration',
        ) as DeclarationReflection,
      ),
    ).toMatchSnapshot();
  });

  test(`should compile declaration with double underscores in name and value`, () => {
    expect(
      declarationPartial(
        global.getMockContext(),
        project.findReflectionByName(
          '__DOUBLE_UNDERSCORES_DECLARATION__',
        ) as DeclarationReflection,
      ),
    ).toMatchSnapshot();
  });

  test(`should compile any function type`, () => {
    expect(
      declarationPartial(
        global.getMockContext(),
        project.findReflectionByName(
          'AnyFunctionType',
        ) as DeclarationReflection,
      ),
    ).toMatchSnapshot();
  });

  test(`should compile function declaration`, () => {
    expect(
      declarationPartial(
        global.getMockContext(),
        project.findReflectionByName(
          'functionDeclaration',
        ) as DeclarationReflection,
      ),
    ).toMatchSnapshot();
  });

  test(`should compile callable declaration`, () => {
    expect(
      declarationPartial(
        global.getMockContext(),
        project.findReflectionByName(
          'callableDeclaration',
        ) as DeclarationReflection,
      ),
    ).toMatchSnapshot();
  });

  test(`should compile indexable declaration`, () => {
    expect(
      declarationPartial(
        global.getMockContext(),
        project.findReflectionByName(
          'indexableDeclaration',
        ) as DeclarationReflection,
      ),
    ).toMatchSnapshot();
  });

  test(`should compile enum delcaration`, () => {
    const props = project.findReflectionByName(
      'EnumDeclarations',
    ) as DeclarationReflection;
    expect(
      declarationPartial(global.getMockContext(), props),
    ).toMatchSnapshot();
  });
});
