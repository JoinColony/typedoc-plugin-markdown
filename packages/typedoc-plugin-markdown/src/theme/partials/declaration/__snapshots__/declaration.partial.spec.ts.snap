// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Declaration Partial: should compile a const with default value 1`] = `
" \`Const\` **stringConstWithDefaultValue**: \`\`\\"hello\\"\`\`

{ sourcesPartial }"
`;

exports[`Declaration Partial: should compile an undefined declaration 1`] = `
" **undefinedNumberDeclaration**: \`number\`

{ sourcesPartial }"
`;

exports[`Declaration Partial: should compile any function type 1`] = `
" **AnyFunctionType**<\`A\`\\\\>: (...\`input\`: \`any\`[]) => [\`A\`]( A )

#### Type parameters

{ typeParameterTablePartial }

#### Type declaration

{ signaturePartial }

{ sourcesPartial }"
`;

exports[`Declaration Partial: should compile callable declaration 1`] = `
" **callableDeclaration**: \`Object\`

#### Call signature

{ signaturePartial }

#### Type declaration

{ propertyTablePartial }

{ sourcesPartial }"
`;

exports[`Declaration Partial: should compile declaration with double underscores in name and value 1`] = `
" \`Const\` **\\\\_\\\\_DOUBLE\\\\_UNDERSCORES\\\\_DECLARATION\\\\_\\\\_**: typeof [\`__DOUBLE_UNDERSCORES_DECLARATION__\`]( __DOUBLE_UNDERSCORES_DECLARATION__ )

{ sourcesPartial }"
`;

exports[`Declaration Partial: should compile enum delcaration 1`] = `
" **EnumDeclarations**: \`Object\`

{ sourcesPartial }"
`;

exports[`Declaration Partial: should compile function declaration 1`] = `
" **functionDeclaration**: (\`someArg\`: \`number\`) => \`boolean\`

{ sourcesPartial }"
`;

exports[`Declaration Partial: should compile indexable declaration 1`] = `
" **indexableDeclaration**: \`Object\`

{ indexSignatureTitlePartial }

#### Type declaration

{ propertyTablePartial }

{ sourcesPartial }"
`;

exports[`Declaration Partial: should compile object literal cast as a const 1`] = `
" \`Const\` **objectLiteralAsConstDeclaration**: \`Object\`

Comments

**\`param\`** Comment for object.

**\`param\`** Comment for Prop1.

**\`param\`** Comment for Prop2.


#### Type declaration

{ propertyTablePartial }

{ sourcesPartial }"
`;

exports[`Declaration Partial: should compile object literal declaration 1`] = `
" \`Const\` **objectLiteralDeclaration**: \`Object\`

**\`param\`** description for valueX

**\`param\`** description for valueZ

**\`param\`** description for valueY


#### Type declaration

{ propertyTablePartial }

{ sourcesPartial }"
`;

exports[`Declaration Partial: should compile type literal declaration 1`] = `
" **typeLiteralDeclaration**: \`Object\`

#### Type declaration

{ propertyTablePartial }

{ sourcesPartial }"
`;
